dnf5/commands/check/check.cpp

```diff
@@ -0,0 +1,315 @@
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "check.hpp"
+
+#include <fmt/format.h>
+#include <libdnf5/utils/bgettext/bgettext-lib.h>
+#include <libdnf5/utils/bgettext/bgettext-mark-domain.h>
+#include <rpm/header.h>
+#include <rpm/rpmcli.h>
+#include <rpm/rpmps.h>
+#include <rpm/rpmtag.h>
+#include <rpm/rpmtd.h>
+#include <rpm/rpmte.h>
+#include <rpm/rpmts.h>
+
+#include <iostream>
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+
+namespace dnf5 {
+
+using namespace libdnf5;
+
+namespace {
+
+enum class ProblemType { MISSING_REQUIRE, CONFLICT, OBSOLETED, DUPLICATE };
+
+// Describes one problem
+struct Problem {
+    ProblemType type;             // type of problem
+    std::string nevra;            // "nevra" causing this problem
+    std::string file_or_provide;  // "file" or "provide" causing this problem
+
+    bool operator<(Problem p) const noexcept {
+        return type < p.type || (type == p.type && (file_or_provide < p.file_or_provide ||
+                                                    (file_or_provide == p.file_or_provide && nevra < p.nevra)));
+    }
+};
+
+// what the program will check
+bool all{true};
+bool dependencies{false};
+bool duplicates{false};
+bool obsoleted{false};
+
+// Contains "provides" to recognize the installonly package.
+// It is initialized with values from the configuration.
+std::vector<std::string> installonly_pkgs_provides;
+
+// Contains a set of problems for each problematic package.
+// The std::map Key is the "nevra" of the problematic package.
+std::map<std::string, std::set<Problem>> problems;
+
+// Contains a list of installed "nevras" for each "name.arch".
+// For "installonly" packages, only one "nevra" is present.
+// Usually contains 1 "nevra" for "name.arch". Multiple "nevras" means "duplicates".
+std::map<std::string, std::vector<std::string>> installed_na_nevras;
+
+
+cli::ArgumentParser::NamedArg * create_option(
+    cli::ArgumentParser & parser, const std::string & name, const std::string & descr, bool & opt_to_modify) {
+    auto option = parser.add_new_named_arg(name);
+    option->set_long_name(name);
+    option->set_description(descr);
+    option->set_parse_hook_func([&opt_to_modify](cli::ArgumentParser::NamedArg *, const char *, const char *) {
+        opt_to_modify = true;
+        return true;
+    });
+    return option;
+}
+
+
+inline bool is_package_installonly(Header h) noexcept {
+    rpmtd td = rpmtdNew();
+    headerGet(h, RPMTAG_PROVIDES, td, 0);
+    while (auto provide = rpmtdNextString(td)) {
+        for (const auto & installonly_pkg_provide : installonly_pkgs_provides) {
+            if (installonly_pkg_provide == provide) {
```

Comment:
I believe that the behavior might differs from libsolv based provide query. On both side it might appear a string with version `kernel < 4`

Reply:
I am not sure about format of the string (`auto provide = rpmtdNextString(td)`), whether it returns name, or  the full string of provide. In most of cases it will not make any difference because we use a lot of provide names and one of them might be not versioned.

Reply:
RPMTAG_PROVIDES returns just the provide names. If you want NEVR strings, use RPMTAG_PROVIDENEVRS instead (and pass HEADERGET_EXT to headerGet() flags)

Reply:
Alternatively you could use the rpmds API so you don't need to care about the low-level header details.

Reply:
`rpm -q --whatprovides 'kernel'` finds packages that provide `kernel = 6.7`

Reply:
I have a good news. I discovered that the current implementation will work according expectations. Our documentation says `List of provide names of packages`. Our config parser is splitting `--setopt=installonlypkgs=' ',"kernel < 5"` to 3 values. It means that there is no requirement of support of versioned provides.
The second part is also perfect -> libsolv job does not support globs `job multiversion provides kernel*` will not match anything. I only discover a minor issue with `dnf5 repoquery --duplicates` where globs are incorrectly supported. I will provide a patch.